[Validator] Update `Type` constraint, add `number`, `finite-float` and `finite-number` validations

guillaume-a opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #50782
| License       | MIT
| Doc PR        | symfony/symfony-docs#18527

Contraint `Type(['float'])` can produce positives with INF or NAN.
This new types car narrow validation and limit validation to finite numbers.

```
#[Assert\Type(['int', 'finite-float'])]
private $value;
```

Thank you for helping me with this one.

guillaume-a committed:
[Validator] Update `Type` constraint, add `number`, `finite-float` and `finite-number` validations

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Allow single integer for the `versions` option of the `Uuid` constraint
  * Allow single constraint to be passed to the `constraints` option of the `When` constraint
+ * Add `Finite` contraint
```
```suggestion
 * Add `Finite` constraint
```

guillaume-a replied:
Oh. Thank you for that :)

stof commented on src/Symfony/Component/Validator/Constraints/Finite.php:
```diff
@@ -0,0 +1,40 @@
+
+    public $message = 'This value should be finite.';
+
+    public function __construct(string $message = null, array $groups = null, mixed $payload = null, array $options = [])
```
to be compatible with the XML and Yaml loaders, the `options` argument must be the first one.

guillaume-a replied:
Made the changes, and also found where ValidatorTests were.
Please tell me if I miss anything.

stof reviewed (changes requested):
This is missing tests

fabpot commented:
@guillaume-a Can you fix fabbot errors?

derrabus commented:
> Contraint `Type(['float'])` can produce false positives.

No, actually that constraint works as advertised because NaN and infinity are valid `float` values.

We should change the PR description.

derrabus commented on src/Symfony/Component/Validator/Tests/Constraints/FiniteValidatorTest.php:
```diff
@@ -0,0 +1,81 @@
+    /**
+     * @dataProvider getValidValues
+     */
+    public function testValidValues($value)
```
```suggestion
    public function testValidValues(mixed $value)
```

derrabus commented on src/Symfony/Component/Validator/Tests/Constraints/FiniteValidatorTest.php:
```diff
@@ -0,0 +1,81 @@
+        $this->assertNoViolation();
+    }
+
+    public static function getValidValues()
```
```suggestion
    public static function getValidValues(): array
```

derrabus commented on src/Symfony/Component/Validator/Tests/Constraints/FiniteValidatorTest.php:
```diff
@@ -0,0 +1,81 @@
+    /**
+     * @dataProvider getInvalidValues
+     */
+    public function testInvalidValues($value, $valueAsString)
```
```suggestion
    public function testInvalidValues(mixed $value, string $valueAsString)
```

derrabus commented on src/Symfony/Component/Validator/Tests/Constraints/FiniteValidatorTest.php:
```diff
@@ -0,0 +1,81 @@
+            ->assertRaised();
+    }
+
+    public static function getInvalidValues()
```
```suggestion
    public static function getInvalidValues(): array
```

derrabus reviewed (commented):
What happens if that validator encounters…
* … a non-numeric string?
* … `null` or an empty string?
* … something completely different lika an array or object?

Can we cover those scenarios with tests?

guillaume-a commented:
Fixed fabbot errors.
Check against `is_numeric()` in the validator (in case of string, array, object `$value`)
Added tests scenarios for all thoses last checks.

`null` was allready checked (considered as valid)
`empty` is now checked and considered as invalid

javiereguiluz commented on src/Symfony/Component/Validator/Constraints/Finite.php:
```diff
@@ -0,0 +1,40 @@
+        self::NOT_FINITE_ERROR => 'NOT_FINITE_ERROR',
+    ];
+
+    public $message = 'This value should be finite.';
```
Just asking if we should make this error message a bit more precise.

`This value should be finite.` -> `This value should be a finite number.`

Maybe we should **not** do this, because the error messages of equivalent constraints are also like this:

`This value should be positive.` (instead of `This value should be a positive number.`)

nicolas-grekas commented:
What about adding new types to the Type constraint instead?
```php
#[Assert\Type(['finite-float'])]
#[Assert\Type(['number'])] # this would be (is_int || is_float) && is_finite
```

guillaume-a commented:
Renamed `Finite` to `IsFinite` to match existing validators `IsTrue` `IsNull` ... 

@nicolas-grekas originally I looked that way, but IMO it will add lots of complexity to `Type` validator since `$types` are considered in a global 'OR' condition. And we should then check againts `is_finite()` only in some cases only which may be hard to deterime.

But i'm open to discussion.

nicolas-grekas commented:
It looks like this would be enough to support my proposal. Not too much complexity IMHO.

```patch
--- a/src/Symfony/Component/Validator/Constraints/TypeValidator.php
+++ b/src/Symfony/Component/Validator/Constraints/TypeValidator.php
@@ -26,9 +26,11 @@ class TypeValidator extends ConstraintValidator
         'int' => 'is_int',
         'integer' => 'is_int',
         'long' => 'is_int',
+        'finite-float' => 'is_float && is_finite',
         'float' => 'is_float',
         'double' => 'is_float',
         'real' => 'is_float',
+        'number' => 'is_int || is_float && is_finite',
         'numeric' => 'is_numeric',
         'string' => 'is_string',
         'scalar' => 'is_scalar',
@@ -69,7 +71,13 @@ class TypeValidator extends ConstraintValidator
 
         foreach ($types as $type) {
             $type = strtolower($type);
-            if (isset(self::VALIDATION_FUNCTIONS[$type]) && self::VALIDATION_FUNCTIONS[$type]($value)) {
+            if (isset(self::VALIDATION_FUNCTIONS[$type])) {
+                match (self::VALIDATION_FUNCTIONS[$type]) {
+                    'finite-float' => \is_float($value) && \is_finite($value),
+                    'number' => \is_int($value) || \is_float($value) && \is_finite($value),
+                    default => self::VALIDATION_FUNCTIONS[$type]($value),
+                };
+
                 return;
             }
```

guillaume-a commented:
I agree. And I love your code, (always forget about `match` and love to re-discover it again and again)

But do we define `finite` as a type, or as an additional condition like `null`, `true` or `false` ?

nicolas-grekas commented:
I'm not sure I'd add `finite`. `finite-float` and `number` look just fine to me. We might want to split number in two possibly: `finite-number` (=int|float&finite) and `number` (=int|float).

guillaume-a commented:
Ok, sounds good to me.
Could we add 

`finite-int` and `finite-float` : to match existing types checking
And a new `number` check with its finite twin `finite-number` ?

nicolas-grekas commented:
`finite-int` doesn't make sense since an int is always finite. But good to me for the others.

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/TypeValidator.php:
```diff
@@ -21,35 +21,38 @@
+      'float' => 'is_float',
+      'double' => 'is_float',
+      'real' => 'is_float',
+      'number' => 'is_int || is_float',
```
please mind the indentation
now I'm wondering: should number be defined as is_int || is_float && !is_nan?

guillaume-a replied:
Sorry, my IDE was not configured properly.
It can make sens for `number` to check againt `is_nan` yes.
And since `is_finite` allready checks that, we can do it only for `number` and not `finite-number`

I'll do it tomorrow.
And i'll update the docs PR too.

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@ CHANGELOG
 
  * Allow single integer for the `versions` option of the `Uuid` constraint
  * Allow single constraint to be passed to the `constraints` option of the `When` constraint
+ * Add `Finite` constraint
```
missing update

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/TypeValidator.php:
```diff
@@ -29,6 +29,9 @@ class TypeValidator extends ConstraintValidator
         'float' => 'is_float',
         'double' => 'is_float',
         'real' => 'is_float',
+        'number' => 'is_int || is_float',
```
```suggestion
        'number' => 'is_int || is_float && !is_nan',
```

ro0NL replied:
for what it's worth, psalm calls it `numeric` and it's a super type of int or float and `numeric-string`

nicolas-grekas replied:
numeric is already a thing for validator, we don't have to follow psalm (we don't already)

number != is-not-a-number

any other opinion on this?

ro0NL replied:
excusi, i missed existing `numeric` below

having both numeric + number is fine :+1: but i'd prefer consistent `finite-*` for all cases

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/TypeValidator.php:
```diff
@@ -69,7 +72,12 @@ public function validate(mixed $value, Constraint $constraint)
+            if (isset(self::VALIDATION_FUNCTIONS[$type]) && match ($type) {
+                'finite-float' => \is_float($value) && is_finite($value),
+                'finite-number' => \is_int($value) || \is_float($value) && is_finite($value),
+                'number' => \is_int($value) || \is_float($value),
```
```suggestion
                'number' => \is_int($value) || \is_float($value) && !is_nan($value),
```

guillaume-a commented:
Can someone explain me why we must write `\is_float` but we must write `is_nan` or `is_finite` without ?
What are the exact rules ?

nicolas-grekas commented:
@guillaume-a check https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048

nicolas-grekas commented:
(can you have a look at the test failure?)

ro0NL commented on src/Symfony/Component/Validator/Constraints/TypeValidator.php:
```diff
@@ -29,6 +29,9 @@ class TypeValidator extends ConstraintValidator
+        'number' => 'is_int || is_float && !is_nan',
+        'finite-float' => 'is_float && is_finite',
+        'finite-number' => 'is_int || is_float && is_finite',
         'numeric' => 'is_numeric',
```
\+ `finite-numeric`?

ro0NL replied:
omfg, ofcourse, how could i miss that.

guillaume-a commented:
> (can you have a look at the test failure?)

Done. (And rebased too.)

fabpot commented:
Thank you @guillaume-a.
